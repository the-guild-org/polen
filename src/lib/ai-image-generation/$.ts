export * as AiImageGeneration from './$$.js'
