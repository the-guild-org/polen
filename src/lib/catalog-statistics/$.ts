export * as CatalogStatistics from './$$.js'
