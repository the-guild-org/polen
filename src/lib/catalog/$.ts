export * as Catalog from './catalog.js'
