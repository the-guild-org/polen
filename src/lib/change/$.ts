export * as Change from './$$.js'
