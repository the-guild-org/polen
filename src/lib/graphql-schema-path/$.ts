export * as GraphQLSchemaPath from './$$.js'
