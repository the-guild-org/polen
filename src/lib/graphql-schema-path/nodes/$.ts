export * as Nodes from './$$.js'
