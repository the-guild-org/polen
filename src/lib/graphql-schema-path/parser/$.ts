export * as Parser from './$$.js'
