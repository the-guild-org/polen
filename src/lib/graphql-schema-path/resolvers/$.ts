export * as Resolvers from './$$.js'
