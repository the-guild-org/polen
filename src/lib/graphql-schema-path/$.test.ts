import { Grafaid } from '#lib/grafaid'
import { Either } from 'effect'
import { buildSchema } from 'graphql'
import { describe, expect, test } from 'vitest'
import { Test } from '../../../tests/unit/helpers/test.js'
import { GraphQLSchemaPath } from './$.js'

describe('.parse', () => {
  // dprint-ignore
  Test.suite<{ input: string }>('valid paths', [
    // Simple types
    { name: 'User type',           input: 'User' },
    { name: 'String scalar',       input: 'String' },
    { name: 'input type',          input: 'CreateUserInput' },

    // Field access
    { name: 'simple field',        input: 'User.name' },
    { name: 'nested field',        input: 'User.posts.title' },

    // Argument access
    { name: 'field argument',      input: 'User.posts$limit' },

    // Resolved types
    { name: 'field resolved type', input: 'User.posts#' },
    { name: 'arg resolved type',   input: 'User.posts$limit#' },

    // Input types
    { name: 'input field',         input: 'CreateUserInput.email' },
    { name: 'nested input',        input: 'CreateUserInput.address.street' },

    // With version
    { name: 'with version',        input: 'v1.0:User' },
    { name: 'multi version',       input: 'v1.0,2.0:User.posts' },
    { name: 'version range',       input: 'v1.0-3.0:Query.users' },
  ], ({ input }) => {
    const result = GraphQLSchemaPath.parse(input)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })

  // dprint-ignore
  Test.suite<{ input: string }>('invalid paths', [
    { name: 'trailing dot',        input: 'User.' },
    { name: 'double dot',          input: 'User..name' },
    { name: 'invalid prefix',      input: 'User$' },
    { name: 'number start',        input: '123User' },
    { name: 'invalid char',        input: 'User!' },
    { name: 'empty string',        input: '' },
  ], ({ input }) => {
    let error: any
    try {
      GraphQLSchemaPath.parse(input)
    } catch (e) {
      error = e
    }
    expect({
      input,
      error: error?.message || error,
    }).toMatchSnapshot()
  })
})

describe('.print', () => {
  // dprint-ignore
  Test.suite<{ input: string }>('round-trip paths', [
    { name: 'simple type',         input: 'User' },
    { name: 'field access',        input: 'User.name' },
    { name: 'nested field',        input: 'User.posts.title' },
    { name: 'with argument',       input: 'User.posts$limit' },
    { name: 'resolved type',       input: 'User.posts#' },
    { name: 'arg resolved',        input: 'User.posts$limit#' },
    { name: 'input field',         input: 'CreateUserInput.email' },
    { name: 'nested input',        input: 'CreateUserInput.address.street' },
    { name: 'with version',        input: 'v1.0:User' },
    { name: 'multi version',       input: 'v1.0,2.0:User.posts' },
    { name: 'version range',       input: 'v1.0-3.0:Query.users' },
  ], ({ input }) => {
    const parsed = GraphQLSchemaPath.parse(input)
    const printed = GraphQLSchemaPath.print(parsed)
    expect({
      input,
      parsed,
      printed,
      roundTripSuccess: printed === input,
    }).toMatchSnapshot()
  })
})

describe('.decodeSync', () => {
  test('decodes string to AST', () => {
    const result = GraphQLSchemaPath.decodeSync('User.posts$limit')
    expect(result._tag).toBe('GraphQLPathRoot')
    expect(result.next._tag).toBe('GraphQLPathSegmentType')
    expect(result.next.name).toBe('User')
    expect(result.next.next?._tag).toBe('GraphQLPathSegmentField')
    expect((result.next.next as any)?.name).toBe('posts')
    expect((result.next.next as any)?.next?._tag).toBe('GraphQLPathSegmentArgument')
    expect((result.next.next as any)?.next?.name).toBe('limit')
  })

  test('round-trips through codec', () => {
    const original = 'User.name'
    const decoded = GraphQLSchemaPath.decodeSync(original)
    const encoded = GraphQLSchemaPath.encodeSync(decoded)
    expect(encoded).toBe(original)
  })

  test('with literal string gets exact type', () => {
    // This test verifies type-level parsing works
    const userPath = GraphQLSchemaPath.decodeSync('User')
    // Type should be ParsePath<'User'> not just GraphQLSchemaPath
    expect(userPath.next.name).toBe('User')
    expect(userPath.next.next).toBeUndefined()

    const complexPath = GraphQLSchemaPath.decodeSync('User.posts$limit')
    expect(complexPath.next.name).toBe('User')
    // TypeScript knows the exact structure from the literal
    expect((complexPath.next.next as any)?.name).toBe('posts')
  })

  test('throws on invalid path', () => {
    expect(() => GraphQLSchemaPath.decodeSync('123Invalid')).toThrow()
    expect(() => GraphQLSchemaPath.decodeSync('')).toThrow()
    expect(() => GraphQLSchemaPath.decodeSync('User..posts')).toThrow()
  })
})

describe('resolver: graphql-schema', () => {
  const schema = buildSchema(`
    type Query {
      user(id: ID!): User
      users(limit: Int): [User!]!
    }

    type User {
      id: ID!
      name: String!
      posts(first: Int): [Post!]!
    }

    type Post {
      id: ID!
      title: String!
      content: String
    }
  `)

  const resolver = GraphQLSchemaPath.Resolvers.GraphqlSchema.create({
    schema,
  })

  const getTypeOrThrow = (name: string) => Grafaid.Schema.Helpers.getTypeOrThrow(schema, name)
  const getFieldedTypeOrThrow = (name: string) => Grafaid.Schema.Helpers.getFieldedTypeOrThrow(schema, name)

  Test.suite<{
    path: string
    expected: { left?: RegExp } | { right?: Grafaid.Schema.TypesLike.Any }
  }>(
    'path resolution',
    // dprint-ignore
    [
      { name: 'resolves type',                    path: 'User',                expected: { right: getTypeOrThrow('User') } },
      { name: 'resolves field',                   path: 'User.name',           expected: { right: getFieldedTypeOrThrow('User').getFields()['name'] } },
      { name: 'resolves field argument',          path: 'User.posts$first',    expected: { right: (getFieldedTypeOrThrow('User').getFields()['posts'] as any).args[0] } },
      { name: 'resolves field resolved type',     path: 'User.posts#',         expected: { right: getTypeOrThrow('Post') } },
      { name: 'resolves argument resolved type',  path: 'User.posts$first#',   expected: { right: getTypeOrThrow('Int') } },
      { name: 'resolves nested field',            path: 'User.posts.title',    expected: { left: /KindMismatch/ } },
      { name: 'resolves query field',             path: 'Query.user',          expected: { right: getFieldedTypeOrThrow('Query').getFields()['user'] } },
      { name: 'resolves query field argument',    path: 'Query.user$id',       expected: { right: (getFieldedTypeOrThrow('Query').getFields()['user'] as any).args[0] } },
      { name: 'fails on non-existent type',       path: 'NonExistent',         expected: { left: /NodeNotFound/ } },
      { name: 'fails on non-existent field',      path: 'User.nonExistent',    expected: { left: /NodeNotFound/ } },
      { name: 'fails on non-existent argument',   path: 'User.name$foo',       expected: { left: /NodeNotFound/ } },
    ],
    ({ path, expected }) => {
      const ast = GraphQLSchemaPath.parse(path)
      const result = resolver(ast)

      if ('right' in expected) {
        expect(result._tag).toBe('Right')
        if (result._tag === 'Right') {
          expect(result.right).toBe(expected.right)
        }
      } else if ('left' in expected) {
        expect(result._tag).toBe('Left')
        if (result._tag === 'Left' && expected.left) {
          // TraversalError has a cause field with the actual StepFailure
          expect(result.left.cause._tag).toMatch(expected.left)
        }
      }
    },
  )
})
