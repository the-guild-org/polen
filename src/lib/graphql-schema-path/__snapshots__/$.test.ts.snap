// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`.parse > invalid paths > 'double dot' 1`] = `
{
  "error": "Expected name but got '.' at position 5",
  "input": "User..name",
}
`;

exports[`.parse > invalid paths > 'empty string' 1`] = `
{
  "error": "Expected name but got 'EOF' at position 0",
  "input": "",
}
`;

exports[`.parse > invalid paths > 'invalid char' 1`] = `
{
  "error": "Unexpected character '!' at position 4",
  "input": "User!",
}
`;

exports[`.parse > invalid paths > 'invalid prefix' 1`] = `
{
  "error": "Unexpected token '$' at position 4",
  "input": "User$",
}
`;

exports[`.parse > invalid paths > 'number start' 1`] = `
{
  "error": "Unexpected character '1' at position 0",
  "input": "123User",
}
`;

exports[`.parse > invalid paths > 'trailing dot' 1`] = `
{
  "error": "Expected name but got 'EOF' at position 5",
  "input": "User.",
}
`;

exports[`.parse > valid paths > 'String scalar' 1`] = `
{
  "input": "String",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "String",
      "next": undefined,
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'User type' 1`] = `
{
  "input": "User",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": undefined,
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'arg resolved type' 1`] = `
{
  "input": "User.posts$limit#",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentArgument",
          "name": "limit",
          "next": {
            "_tag": "GraphQLPathSegmentResolvedType",
          },
        },
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'field argument' 1`] = `
{
  "input": "User.posts$limit",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentArgument",
          "name": "limit",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'field resolved type' 1`] = `
{
  "input": "User.posts#",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentResolvedType",
        },
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'input field' 1`] = `
{
  "input": "CreateUserInput.email",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "CreateUserInput",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "email",
        "next": undefined,
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'input type' 1`] = `
{
  "input": "CreateUserInput",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "CreateUserInput",
      "next": undefined,
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'multi version' 1`] = `
{
  "input": "v1.0,2.0:User.posts",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": undefined,
      },
    },
    "version": "v1.0,2.0",
  },
}
`;

exports[`.parse > valid paths > 'nested field' 1`] = `
{
  "input": "User.posts.title",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentField",
          "name": "title",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'nested input' 1`] = `
{
  "input": "CreateUserInput.address.street",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "CreateUserInput",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "address",
        "next": {
          "_tag": "GraphQLPathSegmentField",
          "name": "street",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'simple field' 1`] = `
{
  "input": "User.name",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "name",
        "next": undefined,
      },
    },
    "version": undefined,
  },
}
`;

exports[`.parse > valid paths > 'version range' 1`] = `
{
  "input": "v1.0-3.0:Query.users",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "Query",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "users",
        "next": undefined,
      },
    },
    "version": "v1.0-3.0",
  },
}
`;

exports[`.parse > valid paths > 'with version' 1`] = `
{
  "input": "v1.0:User",
  "result": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": undefined,
    },
    "version": "v1.0",
  },
}
`;

exports[`.print > round-trip paths > 'arg resolved' 1`] = `
{
  "input": "User.posts$limit#",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentArgument",
          "name": "limit",
          "next": {
            "_tag": "GraphQLPathSegmentResolvedType",
          },
        },
      },
    },
    "version": undefined,
  },
  "printed": "User.posts$limit#",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'field access' 1`] = `
{
  "input": "User.name",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "name",
        "next": undefined,
      },
    },
    "version": undefined,
  },
  "printed": "User.name",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'input field' 1`] = `
{
  "input": "CreateUserInput.email",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "CreateUserInput",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "email",
        "next": undefined,
      },
    },
    "version": undefined,
  },
  "printed": "CreateUserInput.email",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'multi version' 1`] = `
{
  "input": "v1.0,2.0:User.posts",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": undefined,
      },
    },
    "version": "v1.0,2.0",
  },
  "printed": "v1.0,2.0:User.posts",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'nested field' 1`] = `
{
  "input": "User.posts.title",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentField",
          "name": "title",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
  "printed": "User.posts.title",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'nested input' 1`] = `
{
  "input": "CreateUserInput.address.street",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "CreateUserInput",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "address",
        "next": {
          "_tag": "GraphQLPathSegmentField",
          "name": "street",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
  "printed": "CreateUserInput.address.street",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'resolved type' 1`] = `
{
  "input": "User.posts#",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentResolvedType",
        },
      },
    },
    "version": undefined,
  },
  "printed": "User.posts#",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'simple type' 1`] = `
{
  "input": "User",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": undefined,
    },
    "version": undefined,
  },
  "printed": "User",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'version range' 1`] = `
{
  "input": "v1.0-3.0:Query.users",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "Query",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "users",
        "next": undefined,
      },
    },
    "version": "v1.0-3.0",
  },
  "printed": "v1.0-3.0:Query.users",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'with argument' 1`] = `
{
  "input": "User.posts$limit",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": {
        "_tag": "GraphQLPathSegmentField",
        "name": "posts",
        "next": {
          "_tag": "GraphQLPathSegmentArgument",
          "name": "limit",
          "next": undefined,
        },
      },
    },
    "version": undefined,
  },
  "printed": "User.posts$limit",
  "roundTripSuccess": true,
}
`;

exports[`.print > round-trip paths > 'with version' 1`] = `
{
  "input": "v1.0:User",
  "parsed": {
    "_tag": "GraphQLPathRoot",
    "next": {
      "_tag": "GraphQLPathSegmentType",
      "name": "User",
      "next": undefined,
    },
    "version": "v1.0",
  },
  "printed": "v1.0:User",
  "roundTripSuccess": true,
}
`;
